Doc intent for python client versioning

mgrauer opened the pull request:
Fixes #1959.

zachmullen commented on docs/python-client.rst:
```diff
@@ -11,6 +11,16 @@ much easier to upload a large, nested hierarchy of data from a local directory
+Versioning
+----------
+
+The Girder python client is semanticly versioned separately from the Girder
```
semantically*

zachmullen commented on docs/python-client.rst:
```diff
@@ -11,6 +11,16 @@ much easier to upload a large, nested hierarchy of data from a local directory
+The Girder python client is semanticly versioned separately from the Girder
+server, but not fully independently because the client and server are coupled.
+The design intent is that every version of the client should work with every
+version of the server, modulo bugfixes, though this is not an explicit
```
"modulo bugfixes" may not be a great phrase to use for the audience. I think we can remove the phrase here, since the line "though this is not an explicit guarantee" kind of captures that sentiment anyway.

mgrauer commented:
I didn't know about that section, I can point to it for more details but that is beyond the level I expect someone who just wants to use the python client will want to see.  I'll revise accordingly.

After reading the details on the python client versioning, I think my description of it being 'semanticly' versioned was correct! 😸

mgrauer commented:
This is RFR.

msmolens commented on docs/development.rst:
```diff
@@ -388,14 +388,19 @@ recommended process for generating a new release is described here.
+The design intent behind the python client package is to work with as many
+versions of the Girder server as possible; think carefully before breaking this
+compatibility. Whenever the main Girder package is released, the python client
+package should also be versioned and released if is has changed since the last
```
if *it

mgrauer replied:
This was actually a reference to the famous quote `It depends upon what the meaning of the word 'is' is`, so when 'is' has changed...

Naw, good catch.

zachmullen commented on docs/development.rst:
```diff
@@ -388,14 +388,19 @@ recommended process for generating a new release is described here.
+The design intent behind the python client package is to work with as many
+versions of the Girder server as possible; think carefully before breaking this
+compatibility. Whenever the main Girder package is released, the python client
+package should also be versioned and released if it has changed since the last
```
We actually don't do this in practice, and given the independence of the two libraries, maybe we shouldn't prescribe this policy at all, unless you think there's a good reason to do so.

mgrauer commented:
For clarity, I didn't change that block in development.rst other than to add the initial sentence.  It looks like I changed the whole block because I updated the spacing.

That is to say, I didn't write most of what is in the diff in that file, and I haven't been paying attention in detail to how the python client is released, so it isn't like I added that in today based on some perception of how we do things or should do things.  I.e., I'm not prescribing what is written there because I didn't write it and I don't know what we do! 

🍹 

We *should* clarify our policy to ourselves, and then write that here.

Since the design intent of the client is to be maximally compatible with different server versions, and that is now captured here in the docs (in two places), and we don't want to link the versions of client and server (beyond the major version, mentioned in how to come up with the client version number, and which I think should stay) I don't think we need to version the client when we version the server.  This is also supported by the fact that we don't actually do it.

We seem to version the client by roughly the same rules as the server, something like whenever a big enough change occurs or when people ask for it.  That might be a bit vague to put into docs, but it does approximate reality, and half a loaf is better than none.

🥖

mgrauer commented:
Thanks to you both for reading this so carefully.
